src/batch/src/executor/insert.rs

```diff
@@ -149,26 +180,48 @@ impl InsertExecutor {
             let stream_chunk =
                 StreamChunk::new(vec![Op::Insert; cap], columns, vis.into_visibility());
 
-            self.dml_manager
-                .write_chunk(self.table_id, self.table_version_id, stream_chunk)
-                .await
+            let txn_msg = TxnMsg::Data(self.txn_id, stream_chunk);
+
+            #[cfg(debug_assertions)]
+            table_dml_handle.check_txn_msg(&txn_msg);
+
+            write_handle.write_txn_msg(txn_msg).await
         };
 
+        notifiers.push(
+            write_handle
+                .write_txn_msg(TxnMsg::Begin(self.txn_id))
+                .await?,
+        );
+
         #[for_await]
-        for data_chunk in self.child.execute() {
-            let data_chunk = data_chunk?;
+        for data_chunk in self.child.take().unwrap().execute() {
+            let data_chunk = match data_chunk {
+                Ok(data_chunk) => data_chunk,
+                Err(err) => {
+                    write_handle
+                        .write_txn_msg(TxnMsg::Rollback(self.txn_id))
+                        .await?;
+                    self.finished = true;
+                    return Err(err);
+                }
```

Comment:
The data stream can also return an error. For example, an UDF remote service is not available, so we need to try to roll back the transaction.